internal/cmd/io_binary.go

```diff
@@ -0,0 +1,267 @@
+package cmd
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net"
+	"net/url"
+	"os/exec"
+	"sync"
+	"time"
+
+	"github.com/Microsoft/go-winio"
+	"github.com/containerd/containerd/namespaces"
+	"github.com/pkg/errors"
+
+	"github.com/Microsoft/hcsshim/internal/log"
+)
+
+const (
+	binaryPipeFmt         = `\\.\pipe\binary-%s-%s`
+	binaryCmdWaitTimeout  = 10 * time.Second
+	binaryCmdStartTimeout = 10 * time.Second
+)
+
+// NewBinaryIO runs a custom binary process for pluggable shim logging driver.
+//
+// Container's IO will be redirected to the logging driver via named pipes, which are
+// passed as "CONTAINER_STDOUT", "CONTAINER_STDERR" environment variables. The logging
+// driver MUST dial a wait pipe passed via "CONTAINER_WAIT" environment variable AND CLOSE
+// it to indicate that it's ready to consume the IO. For customer's convenience container ID
+// and namespace are also passed via "CONTAINER_ID" and "CONTAINER_NAMESPACE".
+//
+// The path to the logging driver can be provided via a URL's host/path. Additional arguments
+// can be passed to the logger via URL query params
+func NewBinaryIO(ctx context.Context, id string, uri *url.URL) (_ UpstreamIO, err error) {
+	ns, err := namespaces.NamespaceRequired(ctx)
+	if err != nil {
+		ns = namespaces.Default
+	}
+
+	var sout, serr, w io.ReadWriteCloser
+
+	stdoutPipe := fmt.Sprintf(binaryPipeFmt, id, "stdout")
+	sout, err = openNPipe(stdoutPipe)
+	if err != nil {
+		return nil, err
+	}
+
+	stderrPipe := fmt.Sprintf(binaryPipeFmt, id, "stderr")
+	serr, err = openNPipe(stderrPipe)
+	if err != nil {
+		return nil, err
+	}
+
+	waitPipe := fmt.Sprintf(binaryPipeFmt, id, "wait")
+	w, err = openNPipe(waitPipe)
```

Comment:
I think it would be more clear to name these something like 

```
waitPipeName := sprintf 
waitPipe, err := openNPipe
```

so that later when checking to see if the binary started, it's more clear where you're reading from

Reply:
Can we do this for the other pipes too? And change the struct fields in `binaryIO` to reflect that too? 

```
stdinPath, stdoutPath, stderrPath string

stdoutPipe, stderrPipe io.ReadWriteCloser
```

I'm not tied to this idea though, what do you think?